frontend/packages/client/src/components/StepByStep/index.js

```diff
@@ -73,6 +59,71 @@ function StepByStep({
     [refs]
   );
 
+  const [searchParams, setSearchParams] = useSearchParams();
+
+  const goToStep = useCallback(
+    (step, replace = false) => {
+      searchParams.set('step', step + 1);
+      setSearchParams(searchParams, { replace });
+    },
+    [searchParams, setSearchParams]
+  );
+  const tryToGoForward = useCallback(() => {
+    const runPreCheckStepAdvance = () => {
+      if (refs.current) {
+        const runCheckResult = refs.current();
+        if (!runCheckResult) {
+          return false;
+        }
+        refs.current = null;
+      }
+      return true;
+    };
+    const canAdvance = runPreCheckStepAdvance();
+    if (currentStep + 1 < steps.length && canAdvance) {
+      goToStep(currentStep + 1);
+    }
+  }, [currentStep, goToStep, refs, steps.length]);
+  const tryToGoBack = useCallback(() => {
+    if (currentStep - 1 >= 0) {
+      goToStep(currentStep - 1);
+    }
+  }, [currentStep, goToStep]);
+
+  useEffect(() => {
+    const displayStep = searchParams.get('step');
+    const displayStepNum = Number(displayStep);
+    const validStepsVals = Object.values(validSteps);
+    const numValidSteps = validStepsVals.length;
+    const allValidSteps = () =>
+      numValidSteps &&
+      validStepsVals.slice(0, displayStepNum - 1).every((val) => val === true);
+    if (
+      displayStep === null || // first time, need to set step in URL
+      displayStep === undefined || // first time, need to set step in URL
+      displayStepNum <= 0 || // prevent out-of-bounds values
+      displayStepNum > steps.length || // prevent out-of-bounds values
+      (displayStepNum > 1 && !allValidSteps()) || // prevent reaching a subsequent step until prior steps have data
+      (!Object.keys(stepsData).length && numValidSteps) // mismatch
+    ) {
+      setValidSteps({});
+      goToStep(0, true);
+    } else if (displayStepNum - 1 !== currentStep) {
+      // navigated, so update currentStep to match
+      setCurrentStep(displayStepNum - 1);
+    } // else valid & matching, so do nothing
+  }, [
+    currentStep,
+    goToStep,
+    searchParams,
+    steps.length,
+    isStepValid,
+    stepsData,
+    validSteps,
+    setValidSteps,
+    setCurrentStep,
+  ]);
+
```

Comment:
Should this be split into two useEffects? 

- One for checking url is valid (it's a number and it falls under the bounds steps number) and
- A second one to check that all steps are valid (each step has valid data) and allow to move to the next/prior step the url is indicating

I think it's better to use multiple effects to separate concerns, useEffect is designed to handle only one concern.
When trying to use only one effect for multiple purposes, it decreases the readability 😢

Reply:
Updated to split out these two concerns.